packages/interactivity-router/src/head.js

```diff
@@ -0,0 +1,91 @@
+/**
+ * Helper to update only the necessary tags in the head.
+ *
+ * @async
+ * @param {Array} newHead The head elements of the new page.
+ *
+ */
+export const updateHead = async ( newHead ) => {
+	// Helper to get the tag id store in the cache.
+	const getTagId = ( tag ) => tag.id || tag.outerHTML;
+
+	// Map incoming head tags by their content.
+	const newHeadMap = new Map();
+	for ( const child of newHead ) {
+		newHeadMap.set( getTagId( child ), child );
+	}
+
+	const toRemove = [];
+
+	// Detect nodes that should be added or removed.
+	for ( const child of document.head.children ) {
+		const id = getTagId( child );
+		// Always remove styles and links as they might change.
+		if ( child.nodeName === 'LINK' || child.nodeName === 'STYLE' )
+			toRemove.push( child );
```

Comment:
If the stylesheet remains the same, I think it should be left as-is. Otherwise, removing a stylesheet and then adding it right back in will likely cause a performance problem.

Example: https://mousy-citrine-lotus.glitch.me/

Note there is a flash of unstyled content. It also causes layout and style recalculation:

![image](https://github.com/WordPress/gutenberg/assets/134745/d4b8c70f-28d2-409b-bdff-9a43862f7e67)

Reply:
Would be nice to be able to reuse DOM diffing from Preact here. (Not sure if that's possible.)

Reply:
I added a follow-up task to [this list](https://github.com/WordPress/gutenberg/pull/59707#issuecomment-2058832203) to explore this after this experiment is merged.

Reply:
I tried using Preact at first, but if the scripts and styles change order or new elements appear in the head, Preact gets confused and deletes and re-adds everything.

I think it will be simpler and more stable if we control it ourselves.